### ISSUE TITLE
Restructure Array to match List

### ISSUE THREAD

**Issue by mlms13** [reazen/relude#27]
This should be pretty quick, since Array is already a member of `Foldable`, and a whole bunch of List functions were rewritten to come for free from Foldable.

**Comment by arecvlohe** [comment]
I think I can help with this 😅

**Comment by andywhite37** [comment]
@arecvlohe - sounds good.  For some background, @mlms13 refactored `Relude_List` to better take advantage of functions that "you get for free" when you implement certain typeclasses.  E.g. if you implement the `Foldable` typeclass, you get a bunch of functions for free from `Foldable` or `Foldables`, because many convenience functions can just be implemented in terms of the base `Foldable` functions.  When I originally wrote the `Relude_List` and `Relude_Array` I special-cased many of these functions because I was just trying to get through both of them, and we didn't have all the other pieces setup yet.  Sometimes it actually makes sense to special case certain functions for the sake of performance, and not use the freebie implementation, but we can handle that case-by-case.

The general approach I think for `List` and `Array` (and other things) is to alias the functions from `Foldable`/`Foldables`/etc. for convenience, and to give ourselves a placeholder to specialize it in the future if we want to.  This would include all typeclasses, not just `Foldable`, but we can get there over time.

E.g. here are a few examples of functions you get for free.  We are just aliasing them in `List` for conveninece and for future-proofing:

https://github.com/reazen/relude/blob/master/src/Relude_List.re#L279-L298

Looking at the code I just linked, I just saw more of those functions that end with `F` that we wanted to rename to end with `By` :cry: - feel free to update those too.
